File: cotsSource/org.apache.qpid/cpp/src/qpid/broker/HeadersExchange.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
#include "qpid/broker/HeadersExchange.h"
#include "qpid/framing/FieldValue.h"
#include "qpid/framing/reply_exceptions.h"
#include "qpid/log/Statement.h"
#include <algorithm>


using namespace qpid::broker;
using namespace qpid::framing;
using namespace qpid::sys;
namespace _qmf = qmf::org::apache::qpid::broker;

// TODO aconway 2006-09-20: More efficient matching algorithm.
// The current search algorithm really sucks.
// Fieldtables are heavy, maybe use shared_ptr to do handle-body.

using namespace qpid::broker;

namespace {
    const std::string all("all");
    const std::string any("any");
    const std::string x_match("x-match");
    const std::string empty;
}

HeadersExchange::HeadersExchange(const string& _name, Manageable* _parent, Broker* b) :
    Exchange(_name, _parent, b)
{
    if (mgmtExchange != 0)
        mgmtExchange->set_type (typeName);
}

HeadersExchange::HeadersExchange(const std::string& _name, bool _durable,
                                 const FieldTable& _args, Manageable* _parent, Broker* b) :
    Exchange(_name, _durable, _args, _parent, b)
{
    if (mgmtExchange != 0)
        mgmtExchange->set_type (typeName);
}

std::string HeadersExchange::getMatch(const FieldTable* args)
{
    if (!args) {
        throw InternalErrorException(QPID_MSG("No arguments given."));
    }
    FieldTable::ValuePtr what = args->get(x_match);
    if (!what) {
        return empty;
    }
    if (!what->convertsTo<std::string>()) {
        throw InternalErrorException(QPID_MSG("Invalid x-match value binding to headers exchange."));
    }
    return what->get<std::string>();
}

bool HeadersExchange::bind(Queue::shared_ptr queue, const string& bindingKey, const FieldTable* args){
    std::string what = getMatch(args);
    if (what != all && what != any)
        throw InternalErrorException(QPID_MSG("Invalid x-match value binding to headers exchange."));

    Binding::shared_ptr binding (new Binding (bindingKey, queue, this, *args));
    if (bindings.add_unless(binding, MatchArgs(queue, args))) {
        if (mgmtExchange != 0) {
            mgmtExchange->inc_bindingCount();
        }
        routeIVE();
        return true;
    } else {
        return false;
    }
}

bool HeadersExchange::unbind(Queue::shared_ptr queue, const string& bindingKey, const FieldTable*){
    if (bindings.remove_if(MatchKey(queue, bindingKey))) {
        if (mgmtExchange != 0) {
            mgmtExchange->dec_bindingCount();
        }
        return true;
    } else {
        return false;
    }
}


void HeadersExchange::route(Deliverable& msg, const string& /*routingKey*/, const FieldTable* args)
{
    if (!args) {
        //can't match if there were no headers passed in
        if (mgmtExchange != 0) {
            mgmtExchange->inc_msgReceives();
            mgmtExchange->inc_byteReceives(msg.contentSize());
            mgmtExchange->inc_msgDrops();
            mgmtExchange->inc_byteDrops(msg.contentSize());
        }
        return;
    }

    PreRoute pr(msg, this);

    ConstBindingList p = bindings.snapshot();
    BindingList b(new std::vector<boost::shared_ptr<qpid::broker::Exchange::Binding> >);
    if (p.get())
    {
        for (std::vector<Binding::shared_ptr>::const_iterator i = p->begin(); i != p->end(); ++i) {
            if (match((*i)->args, *args)) {
                b->push_back(*i);
            }
        }
    }
    doRoute(msg, b);
}


bool HeadersExchange::isBound(Queue::shared_ptr queue, const string* const, const FieldTable* const args)
{
    Bindings::ConstPtr p = bindings.snapshot();
    if (p.get()){
        for (std::vector<Binding::shared_ptr>::const_iterator i = p->begin(); i != p->end(); ++i) {
            if ( (!args || equal((*i)->args, *args)) && (!queue || (*i)->queue == queue)) {
                return true;
            }
        }
    }
    return false;
}

HeadersExchange::~HeadersExchange() {}

const std::string HeadersExchange::typeName("headers");

namespace 
{

    bool match_values(const FieldValue& bind, const FieldValue& msg) {
        return  bind.empty() || bind == msg;
    }

}


bool HeadersExchange::match(const FieldTable& bind, const FieldTable& msg) {
    typedef FieldTable::ValueMap Map;
    std::string what = getMatch(&bind);
    if (what == all) {
        for (Map::const_iterator i = bind.begin();
             i != bind.end();
             ++i)
        {
            if (i->first != x_match) 
            {
                Map::const_iterator j = msg.find(i->first);
                if (j == msg.end()) return false;
                if (!match_values(*(i->second), *(j->second))) return false;
            }
        }
        return true;
    } else if (what == any) {
        for (Map::const_iterator i = bind.begin();
             i != bind.end();
             ++i)
        {
            if (i->first != x_match) 
            {
                Map::const_iterator j = msg.find(i->first);
                if (j != msg.end()) {
                    if (match_values(*(i->second), *(j->second))) return true;
                }
            }
        }
        return false;
    } else {
        return false;
    }
}

bool HeadersExchange::equal(const FieldTable& a, const FieldTable& b) {
    typedef FieldTable::ValueMap Map;
    for (Map::const_iterator i = a.begin();
         i != a.end();
         ++i)
    {
        Map::const_iterator j = b.find(i->first);
        if (j == b.end()) return false;
        if (!match_values(*(i->second), *(j->second))) return false;
    }
    return true;
}

HeadersExchange::MatchArgs::MatchArgs(Queue::shared_ptr q, const qpid::framing::FieldTable* a) : queue(q), args(a) {}
bool HeadersExchange::MatchArgs::operator()(Exchange::Binding::shared_ptr b)
{
    return b->queue == queue && b->args == *args;
}

HeadersExchange::MatchKey::MatchKey(Queue::shared_ptr q, const std::string& k) : queue(q), key(k) {}

bool HeadersExchange::MatchKey::operator()(Exchange::Binding::shared_ptr b)
{
    return b->queue == queue && b->key == key;
}

File: cotsSource/org.apache.qpid/java/broker/src/main/java/org/apache/qpid/server/exchange/HeadersExchange.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.server.exchange;

import org.apache.log4j.Logger;
import org.apache.qpid.AMQException;
import org.apache.qpid.management.common.mbeans.annotations.MBeanConstructor;
import org.apache.qpid.management.common.mbeans.annotations.MBeanDescription;
import org.apache.qpid.protocol.AMQConstant;
import org.apache.qpid.exchange.ExchangeDefaults;
import org.apache.qpid.framing.AMQShortString;
import org.apache.qpid.framing.AMQTypedValue;
import org.apache.qpid.framing.BasicContentHeaderProperties;
import org.apache.qpid.framing.ContentHeaderBody;
import org.apache.qpid.framing.FieldTable;
import org.apache.qpid.server.queue.AMQQueue;
import org.apache.qpid.server.virtualhost.VirtualHost;
import org.apache.qpid.server.message.InboundMessage;
import org.apache.qpid.server.message.AMQMessageHeader;

import javax.management.JMException;
import javax.management.openmbean.ArrayType;
import javax.management.openmbean.CompositeData;
import javax.management.openmbean.CompositeDataSupport;
import javax.management.openmbean.CompositeType;
import javax.management.openmbean.OpenDataException;
import javax.management.openmbean.OpenType;
import javax.management.openmbean.SimpleType;
import javax.management.openmbean.TabularData;
import javax.management.openmbean.TabularDataSupport;
import javax.management.openmbean.TabularType;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ConcurrentHashMap;

/**
 * An exchange that binds queues based on a set of required headers and header values
 * and routes messages to these queues by matching the headers of the message against
 * those with which the queues were bound.
 * <p/>
 * <pre>
 * The Headers Exchange
 *
 *  Routes messages according to the value/presence of fields in the message header table.
 *  (Basic and JMS content has a content header field called "headers" that is a table of
 *   message header fields).
 *
 *  class = "headers"
 *  routing key is not used
 *
 *  Has the following binding arguments:
 *
 *  the X-match field - if "all", does an AND match (used for GRM), if "any", does an OR match.
 *  other fields prefixed with "X-" are ignored (and generate a console warning message).
 *  a field with no value or empty value indicates a match on presence only.
 *  a field with a value indicates match on field presence and specific value.
 *
 *  Standard instances:
 *
 *  amq.match - pub/sub on field content/value
 *  </pre>
 */
public class HeadersExchange extends AbstractExchange
{

    private static final Logger _logger = Logger.getLogger(HeadersExchange.class);

    public static final ExchangeType<HeadersExchange> TYPE = new ExchangeType<HeadersExchange>()
    {

        public AMQShortString getName()
        {
            return ExchangeDefaults.HEADERS_EXCHANGE_CLASS;
        }

        public Class<HeadersExchange> getExchangeClass()
        {
            return HeadersExchange.class;
        }

        public HeadersExchange newInstance(VirtualHost host, AMQShortString name, boolean durable, int ticket,
                boolean autoDelete) throws AMQException
        {
            HeadersExchange exch = new HeadersExchange();

            exch.initialise(host, name, durable, ticket, autoDelete);
            return exch;
        }

        public AMQShortString getDefaultExchangeName()
        {

            return ExchangeDefaults.HEADERS_EXCHANGE_NAME;
        }
    };


    private final List<Registration> _bindings = new CopyOnWriteArrayList<Registration>();
    private Map<AMQShortString, Registration> _bindingByKey = new ConcurrentHashMap<AMQShortString, Registration>();

    /**
     * HeadersExchangeMBean class implements the management interface for the
     * Header Exchanges.
     */
    @MBeanDescription("Management Bean for Headers Exchange")
    private final class HeadersExchangeMBean extends ExchangeMBean
    {
        @MBeanConstructor("Creates an MBean for AMQ Headers exchange")
        public HeadersExchangeMBean() throws JMException
        {
            super();
            _exchangeType = "headers";
            init();
        }

        /**
         * initialises the OpenType objects.
         */
        protected void init() throws OpenDataException
        {

            _bindingItemTypes = new OpenType[3];
            _bindingItemTypes[0] = SimpleType.INTEGER;
            _bindingItemTypes[1] = SimpleType.STRING;
            _bindingItemTypes[2] = new ArrayType(1, SimpleType.STRING);
            _bindingDataType = new CompositeType("Exchange Binding", "Queue name and header bindings",
                    HEADERS_COMPOSITE_ITEM_NAMES, HEADERS_COMPOSITE_ITEM_DESC, _bindingItemTypes);
            _bindinglistDataType = new TabularType("Exchange Bindings", "List of exchange bindings for " + getName(),
                                                   _bindingDataType, HEADERS_TABULAR_UNIQUE_INDEX);
        }

        public TabularData bindings() throws OpenDataException
        {
            _bindingList = new TabularDataSupport(_bindinglistDataType);
            int count = 1;
            for (Iterator<Registration> itr = _bindings.iterator(); itr.hasNext();)
            {
                Registration registration = itr.next();
                String queueName = registration.queue.getName().toString();

                HeadersBinding headers = registration.binding;
                FieldTable headerMappings = headers.getMappings();
                final List<String> mappingList = new ArrayList<String>();

                headerMappings.processOverElements(new FieldTable.FieldTableElementProcessor()
                {

                    public boolean processElement(String propertyName, AMQTypedValue value)
                    {
                        mappingList.add(propertyName + "=" + value.getValue());
                        return true;
                    }

                    public Object getResult()
                    {
                        return mappingList;
                    }
                });


                Object[] bindingItemValues = {count++, queueName, mappingList.toArray(new String[0])};
                CompositeData bindingData = new CompositeDataSupport(_bindingDataType, HEADERS_COMPOSITE_ITEM_NAMES, bindingItemValues);
                _bindingList.put(bindingData);
            }

            return _bindingList;
        }

        /**
         * Creates bindings. Binding pattern is as follows-
         * <attributename>=<value>,<attributename>=<value>,...
         * @param queueName
         * @param binding
         * @throws javax.management.JMException
         */
        public void createNewBinding(String queueName, String binding) throws JMException
        {
            AMQQueue queue = getQueueRegistry().getQueue(new AMQShortString(queueName));

            if (queue == null)
            {
                throw new JMException("Queue \"" + queueName + "\" is not registered with the exchange.");
            }

            String[] bindings = binding.split(",");
            FieldTable bindingMap = new FieldTable();
            for (int i = 0; i < bindings.length; i++)
            {
                String[] keyAndValue = bindings[i].split("=");
                if (keyAndValue == null || keyAndValue.length == 0 || keyAndValue.length > 2)
                {
                    throw new JMException("Format for headers binding should be \"<attribute1>=<value1>,<attribute2>=<value2>\" ");
                }

                if(keyAndValue.length ==1)
                {
                    //no value was given, only a key. Use an empty value
                    //to signal match on key presence alone
                    bindingMap.setString(keyAndValue[0], "");
                }
                else
                {
                    bindingMap.setString(keyAndValue[0], keyAndValue[1]);
                }
            }

            _bindings.add(new Registration(new HeadersBinding(bindingMap), queue, new AMQShortString(binding)));
        }

    } // End of MBean class

    public AMQShortString getType()
    {
        return ExchangeDefaults.HEADERS_EXCHANGE_CLASS;
    }

    public void registerQueue(AMQShortString routingKey, AMQQueue queue, FieldTable args) throws AMQException
    {
        _logger.debug("Exchange " + getName() + ": Binding " + queue.getName() + " with " + args);

        Registration registration = new Registration(new HeadersBinding(args), queue, routingKey);
        _bindings.add(registration);

    }

    public void deregisterQueue(AMQShortString routingKey, AMQQueue queue, FieldTable args) throws AMQException
    {
        _logger.debug("Exchange " + getName() + ": Unbinding " + queue.getName());

        if(!_bindings.remove(new Registration(args == null ? null : new HeadersBinding(args), queue, routingKey)))
        {
            throw new AMQException(AMQConstant.NOT_FOUND, "Queue " + queue + " was not registered with exchange " + this.getName()
                                   + " with headers args " + args);
        }
    }

    public ArrayList<AMQQueue> route(InboundMessage payload)
    {
        AMQMessageHeader header = payload.getMessageHeader();
        if (_logger.isDebugEnabled())
        {
            _logger.debug("Exchange " + getName() + ": routing message with headers " + header);
        }
        boolean routed = false;
        ArrayList<AMQQueue> queues = new ArrayList<AMQQueue>();
        for (Registration e : _bindings)
        {

            if (e.binding.matches(header))
            {
                if (_logger.isDebugEnabled())
                {
                    _logger.debug("Exchange " + getName() + ": delivering message with headers " +
                                  header + " to " + e.queue.getName());
                }
                queues.add(e.queue);

                routed = true;
            }
        }
        return queues;
    }

    public boolean isBound(AMQShortString routingKey, FieldTable arguments, AMQQueue queue)
    {
        //fixme isBound here should take the arguements in to consideration.
        return isBound(routingKey, queue);
    }

    public boolean isBound(AMQShortString routingKey, AMQQueue queue)
    {
        return isBound(queue);
    }

    public boolean isBound(AMQShortString routingKey)
    {
        return hasBindings();
    }

    public boolean isBound(AMQQueue queue)
    {
        for (Registration r : _bindings)
        {
            if (r.queue.equals(queue))
            {
                return true;
            }
        }
        return false;
    }

    public boolean hasBindings()
    {
        return !_bindings.isEmpty();
    }

    protected FieldTable getHeaders(ContentHeaderBody contentHeaderFrame)
    {
        //what if the content type is not 'basic'? 'file' and 'stream' content classes also define headers,
        //but these are not yet implemented.
        return ((BasicContentHeaderProperties) contentHeaderFrame.properties).getHeaders();
    }

    protected ExchangeMBean createMBean() throws JMException
    {
        return new HeadersExchangeMBean();
    }

    public Map<AMQShortString, List<AMQQueue>> getBindings()
    {
        return null;
    }

    public Logger getLogger()
    {
        return _logger;
    }


    private static class Registration
    {
        private final HeadersBinding binding;
        private final AMQQueue queue;
        private final AMQShortString routingKey;

        Registration(HeadersBinding binding, AMQQueue queue, AMQShortString routingKey)
        {
            this.binding = binding;
            this.queue = queue;
            this.routingKey = routingKey;
        }

        public int hashCode()
        {
            int queueHash = queue.hashCode();
            int routingHash = routingKey == null ? 0 : routingKey.hashCode();
            return queueHash + routingHash;
        }

        public boolean equals(Object o)
        {
            return o instanceof Registration
                   && ((Registration) o).queue.equals(queue)
                   && (routingKey == null ? ((Registration)o).routingKey == null
                                          : routingKey.equals(((Registration)o).routingKey));
        }
    }
}
